Fix: Don't step into very first group automatically

cooperka opened the pull request:
#### Changes

- Fixes https://github.com/opendatakit/collect/issues/2871
    - This is a long-standing bug, explicitly verified in v1.18.0 but likely around for much longer than that. The logic this modifies was last updated about [7 years ago](https://github.com/opendatakit/collect/commit/920b6a28dbc35208a0bd232100e3fb2f0946af84). High-risk.

- Fixes #2936
    - Simple, seemingly safe change: neither a visible group nor a repeat group should be rendered inside someone else's parent.

These two separate issues are highly related so I want to make sure they're tested together.

#### What has been done to verify that this works as intended?

Test on the updated [basic.xml](https://github.com/opendatakit/collect/files/2861270/basic.xml.txt). The form has an initial group with some questions, followed by questions outside of the group. All 4 questions are expected to appear on the jump screen, but this bug caused only the grouped questions to appear.

Also test with `groupTestNav` (see instructions on #2936).

This should also be tested on all sorts of other types of forms, specifically those with a group as the very first element, including field-lists.

#### Why is this the best possible solution? Were any other approaches considered?

Stepping into a non-visible group doesn't make sense, and I'm not sure why the code was written this way. Removing this logic fixed the bug.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Users will now be able to see all questions on the jump screen, as they would expect.

Regression risks include unexpected edge cases with any form that starts with a group as the very first element.

#### Do we need any specific form for testing your changes? If so, please attach one.

Linked above.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

No

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

cooperka committed:
Don't step into first group of form automatically

cooperka commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -565,7 +562,7 @@ private void refreshView(boolean isGoingUp) {
                 TreeReference curGroup = (visibleGroupRef == null) ? contextGroupRef : visibleGroupRef;
 
-                if (!curGroup.isParentOf(currentRef, false)) {
+                if (curGroup != null && !curGroup.isParentOf(currentRef, false)) {
```
Explanation: Check for `null` here because now `contextGroupRef` may not have been set.

cooperka committed:
Remove unused method

grzesiek2010 commented:
Uff good to know it's not caused by the recent changes. Definitely let's wait with merging.

grzesiek2010 commented:
This pr should be very carefully tested to avoid regression in other cases. Whoever will take it @mmarciniak90 @kkrawczy123, please test various forms with groups, nested groups etc.

mmarciniak90 commented:
@cooperka 
I compared behaviors on v1.20 and app with the fix on mentioned forms:
* basic
* RepeatSimple
* repeatTest
* TestRepeat
* nRepeats
* nRepeatsWithoutID
* nested-repeat-complex
* nested-repeats
* fewRepeatGroups 
* one repeat

During checking [repeat_group_form.xml.txt](https://github.com/opendatakit/collect/files/2926662/repeat_group_form.xml.txt) form, I noticed unexpected dialog.
I was on the last view with `Save Form and Exit` and I clicked on `Jump` icon. 

|Android 7.0 | Android 4.4|
|---|---|
|![screenshot_20190304-151623](https://user-images.githubusercontent.com/3907554/53738826-8e973e00-3e90-11e9-9942-b05848e1706b.png)|![screenshot_2019-03-04-16-18-57](https://user-images.githubusercontent.com/3907554/53738958-e9309a00-3e90-11e9-8855-4a53a7c6c5c8.png)|

zestyping commented:
I believe this is missing another `contextGroupRef != null` check on line 609:

                        // Don't render other groups' instances.
                        if (!contextGroupRef.isParentOf(currentRef, false)) {
                            break;
                        }

Pressing the Up button twice in the Repeat Group form causes a null pointer exception at this spot.

[repeat_group_form.xml.txt](https://github.com/opendatakit/collect/files/2935789/repeat_group_form.xml.txt)

zestyping commented:
Otherwise this seems okay.

zestyping commented:
Incidentally, playing with this form makes me more convinced that the breadcrumb trail should say "1. Alice", not "1 > Alice".

When I start here:
![2019-03-06 11-00-52](https://user-images.githubusercontent.com/236086/53876699-38271e00-3fbc-11e9-8bce-0954001735da.png)

and click "1. Hi 'abc'", I get here:
![2019-03-06 11-01-08](https://user-images.githubusercontent.com/236086/53876686-32313d00-3fbc-11e9-9b47-ef67767deafd.png)

If I then click Up, I go back to here:
![2019-03-06 11-00-52](https://user-images.githubusercontent.com/236086/53876699-38271e00-3fbc-11e9-8bce-0954001735da.png)

So the navigation very much conveys that this is descending one level and ascending one level.  The number of ">" should indicate the number of levels; descending a level should add one ">" and going up one level should remove one ">".

Or, in short: **One click down, one click up** => **one level down, one level up** => **one ">"**

cooperka commented:
Thank you both, I'll do my best to address these issues by early next week.

@zestyping I completely agree that it should be `1.` instead of `1 >` but that requires a bit of refactor which is somewhat risky. Behind the scenes FormHierarchy is actually skipping past the group containing `1` and showing you only the group containing `1 > Hi`. When you go "up" it skips back out past the group containing `1` again. There are really two nested groups in the form, but it **feels** like there's only one, for efficiency of navigation.

See https://github.com/opendatakit/collect/pull/2764/files#diff-c4569759abab98c20df2aa91fe129bc8R520 for more context.

zestyping commented:
Thanks!

cooperka committed:
Merge branch 'master' into fix-last-q

cooperka committed:
Check for null at the only other place contextGroupRef is used

cooperka commented:
Fixed! @opendatakit-bot label "needs testing"

cooperka committed:
Also skip unrelated repeat groups within visible groups

Same logic here as above; neither a visible group nor a repeat group
should be rendered inside someone else's parent.

cooperka commented:
@grzesiek2010 I also included the simple fix for https://github.com/opendatakit/collect/issues/2936 in here. These two fixes are highly related so I want to make sure they're tested together.

mmarciniak90 commented:
I re-tested issue with success.
I did not notice other unexpected behaviors.

Verified cases (Android 4.2, 4.4, 5.1, 6.0, 7.0, 8.1):
* Case with `groupTestNav` form - 2936
* Case with `basic` form - 2871 
* correct behavior in  `repeat_group_form` form

Forms used to compare with v1.20 (Android 7.0, 4.4)
* nested-index-repeat
* nested-repeat
* nested-repeat-complex
* nRepeats
* one repeat
* repeat_inside_regular
* RepeatGroupAndGroup
* RepeatGroupAndQuestion
* RepeatGroupNew
* RepeatSimple
* repeatTest
* TestRepeat

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
